### PR TITLE
Db auto decay

### PR DESCRIPTION
**Pull request by shin202j**
Final, hopefully

### CODE REVIEW

File: Sources/DatabaseEndpoints.swift [hunk 1]
```diff
@@ -33,7 +33,8 @@ public class LXDataBaseEndpoint: LXEndpoint {
     public var entryFormatter: LXEntryFormatter
     /// This Endpoint requires a newline character appended to each serialized Log Entry string.
     public let requiresNewlines: Bool = true
-
+    public var dataBin:[String] = []
```
Comment (bunnyhugdev) [inline Sources/DatabaseEndpoints.swift]: Make this a maxTimestamp or something like that. You can use it to mark as sent anything up to and including this timestamp.

File: Sources/DatabaseEndpoints.swift [hunk 2]
```diff
@@ -94,23 +95,66 @@ public class LXDataBaseEndpoint: LXEndpoint {
...
             let flagDown = try managedContext.fetch(fetchRequest)
-            let objectUpdate = flagDown[0] as! NSManagedObject
-            objectUpdate.setValue(true, forKey: "sent")
-            
-            saveContext(managedContext: managedContext)
+            if (flagDown.count > 0){
+                for i in 0...flagDown.count - 1{
+                    let objectUpdate = flagDown[i] as! NSManagedObject
+                    dataBin[i] = "\(objectUpdate.value(forKey: "timeStamp") ?? "empty")"
+                    resultString = "\(resultString) \(objectUpdate.value(forKey: "timeStamp") ?? "empty")"
```
Comment (bunnyhugdev) [inline Sources/DatabaseEndpoints.swift]: Lets create an array of strings for each message.
Also only need message attribute - don't need timestamp or sent.

File: Sources/DatabaseEndpoints.swift [hunk 3]
```diff
@@ -94,23 +95,66 @@ public class LXDataBaseEndpoint: LXEndpoint {
...
+                    resultString = "\(resultString) \(objectUpdate.value(forKey: "message") ?? "empty")"
+                    resultString = "\(resultString) \(objectUpdate.value(forKey: "sent") ?? "empty")"
+                    resultString += "\n"
+                }
+                saveContext(managedContext: managedContext)
```
Comment (bunnyhugdev) [inline Sources/DatabaseEndpoints.swift]: Probably don't need a save context as we aren't changing anything.

File: Sources/DatabaseEndpoints.swift [hunk 4]
```diff
@@ -94,23 +95,66 @@ public class LXDataBaseEndpoint: LXEndpoint {
...
+            }
+            else{
+                return "There is no new logs"
```
Comment (bunnyhugdev) [inline Sources/DatabaseEndpoints.swift:115]: return []

File: Sources/DatabaseEndpoints.swift [hunk 5]
```diff
@@ -94,23 +95,66 @@ public class LXDataBaseEndpoint: LXEndpoint {
...
         }
+        dataBin = []
+        saveContext(managedContext: managedContext)
     }
-  
+    
+    public func getLogsFromDatabase() -> Data {
+        let str = updateData()
+        let data = str.data(using: String.Encoding.utf8)
+        return data!
```
Comment (bunnyhugdev) [inline Sources/DatabaseEndpoints.swift:158]: I'd return Array of strings here instead of Data

File: Sources/HTTPEndpoints.swift [hunk 6]
```diff
@@ -217,6 +217,27 @@ public class LXHTTPEndpoint: LXEndpoint {
 
         self.timer.fire()
     }
+
```
Comment (bunnyhugdev) [inline Sources/HTTPEndpoints.swift:220]: This file shouldn't change here.

File: Sources/LogKit.swift [hunk 7]
```diff
@@ -244,10 +243,51 @@ internal extension NSCalendar {
     @objc static func critical(message: String) {
         LogKit.logger.critical(message: message, functionName: getFunctionInfo())
     }
+  
+    @objc static func pushToServer(url: NSURL, completion: @escaping (Bool, String) -> Void) {
+        let resultLogs = LogKit.logger.getLogsData()
+        let marking = LXDataBaseEndpoint()
```
Comment (bunnyhugdev) [inline Sources/LogKit.swift]: No LXDatabaseEndpoint here

File: Sources/LogKit.swift [hunk 8]
```diff
@@ -244,10 +243,51 @@ internal extension NSCalendar {
...
+        //create the session object
+        let session = URLSession.shared
+        
+        //now create the Request object using the url object
+        var request = URLRequest(url: url as URL)
+        request.httpMethod = "POST" //set http method as POST
+        request.httpBody = resultLogs
```
Comment (bunnyhugdev) [inline Sources/LogKit.swift:255]: since resultLogs is array, join them all with \n

File: Sources/LogKit.swift [hunk 9]
```diff
@@ -244,10 +243,51 @@ internal extension NSCalendar {
...
+        
+        //HTTP Headers
+        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
```
Comment (bunnyhugdev) [inline Sources/LogKit.swift]: Change to text/plain for Content-Type

File: Sources/LogKit.swift [hunk 10]
```diff
@@ -244,10 +243,51 @@ internal extension NSCalendar {
...
+        let task = session.dataTask(with: request, completionHandler: { data, response, error in
+            
+            guard error == nil else {
+                completion(false, error as! String)
+                return
+            }
+            
+            do {
+                //create json object from data
+                guard let json = try JSONSerialization.jsonObject(with: data!, options: .mutableContainers) as? [String: Any] else {
```
Comment (bunnyhugdev) [inline Sources/LogKit.swift:271]: Wont need to serialize to json

File: Sources/Loggers.swift [hunk 11]
```diff
@@ -37,6 +37,12 @@ public final class LXLogger{
         self.init(endpoints: [LXDataBaseEndpoint()])
     }
 
+    func getLogsData() -> Data {
+        for endpoint in self.endpoints {
+            return endpoint.getLogsFromDatabase()
```
Comment (bunnyhugdev) [inline Sources/Loggers.swift]: logs.append( endpoint.getLogsFromDatabase() )

File: Sources/Loggers.swift [hunk 12]
```diff
@@ -37,6 +37,12 @@ public final class LXLogger{
...
+        }
+        return NSData() as Data
```
Comment (bunnyhugdev) [inline Sources/Loggers.swift]: return logs;

File: Sources/LogKit.swift (hunk 10 above)
Comment (omeryounus) [reply Sources/LogKit.swift:271]: So this handles json response from server. 
{
  "bytes_written": 19,
  "lines_written": 2,
  "status": "OK"
}

**Review by bunnyhugdev — APPROVED** [review]

**Pull request merged**